### DISCUSSION TITLE
Limit repaint area after rerendering

### DISCUSSION THREAD

**Discussion by hrdl-github** [xournalpp/xournalpp#4494] [General] [answered]
While investigating some performance issues in combination with electrophoretic displays I stumbled upon this call: https://github.com/xournalpp/xournalpp/blob/c9ac931bb24a53c332d3bda0f61ce768936d683d/src/core/control/jobs/RenderJob.cpp#L87 

This line effectively repaints the entire Xournal widget. For a lot of modifications (e.g. for most basic strokes and shapes) this is not needed and removing this call does not affect the visual output while greatly speeding up drawing on those displays. However, operations such as standard erasure or stroke-based erasure don't get displayed in this case until another repaint happens.

Do you have any suggestions how the number of full repaints can be reduced to accommodate this scenario? I've tried limiting the repaint area to the bounding box of the union of rectangles to be rerendered, which has still some bugs and causes the application to crash. I will report back once I've fixed this. Another idea would be to make a full repaint only if an element or operation indicates that a repaint is required.

**Comment by Technius** [comment]
/CC @bhennion since he has been working on the rendering code lately.

Thanks for investigating the performance. We're currently in the process of refactoring the way elements are rendered (this is quite messy, see #4051), but hopefully this should be easier in the future.

> However, operations such as standard erasure or stroke-based erasure don't get displayed in this case until another repaint happens.

These are probably bugs; we may be missing calls to `RepaintHandler` methods in several places.

> Do you have any suggestions how the number of full repaints can be reduced to accommodate this scenario? I've tried limiting the repaint area to the bounding box of the union of rectangles to be rerendered, which has still some bugs and causes the application to crash.

I think what you tried is the approach that we'll want to take in the future, it'll just take some time and effort to get there. In the short term, one optimization that I think shouldn't require too many modifications is to repaint only the bounds of the page on the canvas (e.g., via `RepaintHandler::repaintPage()`).

**Comment by bhennion** [comment]
Indeed, the rerendering job should only repaint the relevant portion of screen.
Simply calling `XojPageView::repaintPage()` if `rerenderComplete == true`, and `XojPageView::flagDirtyRegion(Range(rect))` in `RenderJob::rerenderRectangle()` seems reasonnable to me.

**Reply by hrdl-github** [reply]
Thanks for your suggestions. I will try to get a better understanding of the current and planned rendering and implement what you suggested sometime this week.

**Comment by hrdl-github** [comment]
I have implemented what  you suggested (https://github.com/xournalpp/xournalpp/pull/4502) and will keep testing this within the next few days. I think there is still room for improvement: at the moment finishing a stroke triggers a repaint, which is apparently not necessary for regular strokes and other "purely constructive" operations. For erasing, undoing, and redoing this is required at the moment. Do you have any thoughts on this? Most of these questions are probably not relevant for regular devices, but for devices like the one I am working it makes all the difference.

**Reply by hrdl-github** [reply]
I am using this approach for now: https://github.com/xournalpp/xournalpp/commit/0e9edbc6a0537f93481a8fded01ff625d1d379e8
Things that I believe are missing include rendering changes induced by the movement between the last motion event and the button release event as well as being more conservative with repainting, so e.g. filled objects are always repainted. This also needs more rigorous testing with different types of erasing and line types.

Are you interested in integrating something like this into xournalpp once polished?

**Comment by hrdl-github** [comment, marked as answer]
https://github.com/xournalpp/xournalpp/pull/4502 was merged into master.
